### PR TITLE
Filtering alerts before distribution

### PR DESCRIPTION
**Pull request by cAbhi15**
Linked to issue(s): #203 

## What changes were proposed in this pull request?
- Add xml based filter before alert distribution

## How was this patch tested?
Manually, CI

### CODE REVIEW

**Comment by JulienPeloton** [comment]
Interesting, thanks @cAbhi15 for bringing here this technique. 

I think for key/value based filters, this would work fine. Especially it seems there is also an opportunity to integrate that easily into a web frontend (users would just select what they want from a web interface, and the xml file would be created on-the-fly in the cluster).

**Comment by JulienPeloton** [comment]
I think once you are done with this, I will use it for the `raw2science` service - at least for the beginning. In the meantime, I will develop the python based technique to allow generic user defined functions to run as well (anything beyond the simple key/value).

File: bin/distribution_test.py [hunk 1]
```diff
@@ -53,9 +53,10 @@ def main():
     df = decode_kafka_df(df_kafka, args.distribution_schema)
 
     # Print to console
-    cols = ["objectId", "candid", "candidate_ra", "candidate_dec", "simbadType"]
-    cols = ["struct." + c for c in cols]
-    df = df.select(cols)
+    # cols = ["objectId", "candid", "candidate_ra", "candidate_dec", "simbadType"]
+    # cols = ["struct." + c for c in cols]
+    # df = df.select(cols)
```
Comment (JulienPeloton) [inline bin/distribution_test.py]: Remove commented code

File: fink_broker/filters.py [hunk 2]
```diff
@@ -70,6 +73,276 @@ def keep_alert_based_on(nbad: Any, rb: Any, magdiff: Any) -> pd.Series:
     return pd.Series(mask)
 
 
+def parse_xml_rules(rules_xml: str, df_cols: list) -> Tuple[list, list]:
```
Comment (JulienPeloton) [inline fink_broker/filters.py]: I think I get the logic here, but the cognitive complexity of this function is huge (48 - given that the threshold is around 15). Is there any possibility to make it more readable and reduce complexity? 

To learn more about cognitive complexity see e.g. https://www.sonarsource.com/docs/CognitiveComplexity.pdf

File: fink_broker/filters.py [hunk 3]
```diff
@@ -70,6 +73,276 @@ def keep_alert_based_on(nbad: Any, rb: Any, magdiff: Any) -> pd.Series:
...
+    # if the xml file doesn't exist or is empty
+    if not os.path.isfile(rules_xml) or os.path.getsize(rules_xml) <= 0:
+        print("invalid xml: does not exist or is empty")
+        return [],[]
+
+    # parse the xml and make an element tree
+    tree = ET.parse(rules_xml)
+    root = tree.getroot()
+
+    #------------------------------------------------------#
```
Comment (JulienPeloton) [inline fink_broker/filters.py]: Block comment should start with `#  ` (with a space after the hashtag)

File: fink_broker/filters.py [hunk 4]
```diff
@@ -70,6 +73,276 @@ def keep_alert_based_on(nbad: Any, rb: Any, magdiff: Any) -> pd.Series:
...
+            if col in df_cols:
+                cols_to_select.append(col)
+            else:
+                print("Error in Select: invalid column: {}".format(col))
+                return [],[]
+
+    # remove duplicates
+    cols_to_select = list(dict.fromkeys(cols_to_select))
+
+    #------------------------------------------------------#
```
Comment (JulienPeloton) [inline fink_broker/filters.py]: Block comment should start with `#  ` (with a space after the hashtag)

File: fink_broker/filters.py [hunk 5]
```diff
@@ -70,6 +73,276 @@ def keep_alert_based_on(nbad: Any, rb: Any, magdiff: Any) -> pd.Series:
...
+                if col in df_cols:
+                    cols_to_drop.append(col)
+                else:
+                    print("Error in Drop: invalid column: {}".format(col))
+                    return [],[]
+
+        # remove duplicates
+        cols_to_drop = list(dict.fromkeys(cols_to_drop))
+
+    #------------------------------------------------------#
```
Comment (JulienPeloton) [inline fink_broker/filters.py]: Unnecessary line.

File: fink_broker/filters.py [hunk 6]
```diff
@@ -70,6 +73,276 @@ def keep_alert_based_on(nbad: Any, rb: Any, magdiff: Any) -> pd.Series:
...
+
+    # obtain columns to distribute
+    cols_to_distribute = [e for e in cols_to_select if e not in cols_to_drop]
+
+    #------------------------------------------------------#
```
Comment (JulienPeloton) [inline fink_broker/filters.py]: Block comment should start with `#  ` (with a space after the hashtag)

File: fink_broker/filters.py [hunk 7]
```diff
@@ -70,6 +73,276 @@ def keep_alert_based_on(nbad: Any, rb: Any, magdiff: Any) -> pd.Series:
...
+                    rule = col + attrib['operator'] + attrib['value']
+                    rules_list.append(rule)
+                else:
+                    print("Error in Filter: invalid column: {}".format(col))
+                    return cols_to_distribute, []
+
+        # remove duplicate rules
+        rules_list = list(dict.fromkeys(rules_list))
+
+    #------------------------------------------------------#
```
Comment (JulienPeloton) [inline fink_broker/filters.py]: Block comment should start with `#  ` (with a space after the hashtag)

File: fink_broker/filters.py [hunk 8]
```diff
@@ -70,6 +73,276 @@ def keep_alert_based_on(nbad: Any, rb: Any, magdiff: Any) -> pd.Series:
...
+
+    # Obtain the Filtered DataFrame:
+    # Select cols to distribute
+    df_filtered = df.select(cols_to_distribute)
+
+    # Apply filters
+    for rule in rules_list:
+        df_filtered = df_filtered.filter(rule)
+
+    return df_filtered.cache()
```
Comment (JulienPeloton) [inline fink_broker/filters.py]: How did you make sure that caching too frequently is not an issue?

File: bin/distribute.py [hunk 9]
```diff
@@ -85,10 +91,10 @@ def main():
...
             .format("kafka")\
             .option("kafka.bootstrap.servers", "localhost:9093")\
-            .option("topic", "distribution_test")\
+            .option("topic", "fink_outstream")\
             .save()
 
-        # Update the status column in Hbase
+        # Update the status in Hbase and commit checkpoint to file
         update_status_in_hbase(df, args.science_db_name, "objectId",
                 args.checkpointpath_dist, max_timestamp)
```
[the remark this answers is not in the document]
Comment (cAbhi15) [reply bin/distribute.py]: fixed in b76ac2c

File: bin/distribution_test.py (hunk 1 above)
Comment (cAbhi15) [reply bin/distribution_test.py]: fixed in 5aa4e20

File: fink_broker/filters.py (hunk 2 above)
Comment (cAbhi15) [reply fink_broker/filters.py]: Thanks @JulienPeloton for the resource on cognitive complexity.
Is it any good now (cd4fee7) ?

File: fink_broker/filters.py (hunk 8 above)
Comment (cAbhi15) [reply fink_broker/filters.py]: will this (bbc199f) do the trick ?

File: fink_broker/filters.py (hunk 2 above)
Comment (JulienPeloton) [reply fink_broker/filters.py]: Thanks @cAbhi15, this is great! Could you just move the inner functions outside of `parse_xml_rules`? They do not need to be inside, and it reduces the readability.

File: fink_broker/filters.py (hunk 8 above)
Comment (JulienPeloton) [reply fink_broker/filters.py]: I will make resource monitoring later on, but it seems a good trick.

File: fink_broker/filters.py (hunk 2 above)
Comment (cAbhi15) [reply fink_broker/filters.py]: done in a58cb3b

Comment (JulienPeloton) [reply fink_broker/filters.py]: Great thanks!

**Comment by JulienPeloton** [comment]
Thanks @cAbhi15 for this great contribution! Merging.

**Pull request merged**